lang/LangPrimSource/PyrFilePrim.cpp

```diff
@@ -1326,6 +1326,67 @@ int prPipeOpen(struct VMGlobals *g, int numArgsPushed)
 	return errNone;
 }
 
+int prPipeOpenArgv(struct VMGlobals *g, int numArgsPushed)
+{
+	PyrSlot *callerSlot, *argsSlot, *modeSlot;
+	char mode[12];
+
+	callerSlot = g->sp - 2;
+	argsSlot = g->sp - 1;
+	modeSlot = g->sp;
+
+#ifdef SC_IPHONE
+	SetInt(callerSlot, 0);
+	return errNone;
+#endif
+
+	if (NotObj(argsSlot)) return errWrongType;
+
+	if (NotObj(modeSlot) || !isKindOf(slotRawObject(modeSlot), class_string))
+		return errWrongType;
+	if (slotRawObject(modeSlot)->size > 11) return errFailed;
+
+	PyrFile *pfile = reinterpret_cast<PyrFile*>(slotRawObject(callerSlot));
+
+	PyrObject *argsColl = slotRawObject(argsSlot);
+	if (!(slotRawInt(&argsColl->classptr->classFlags) & classHasIndexableInstances))
+		return errNotAnIndexableObject;
+
+	if( argsColl->size < 1)
+		return errFailed;
+
+	memcpy(mode, slotRawString(modeSlot)->s, slotRawObject(modeSlot)->size);
+	mode[slotRawString(modeSlot)->size] = 0;
+
+	std::vector<char *> argv (argsColl->size + 1);
+	argv[argsColl->size] = nullptr;
+
+	for (int i=0; i<argsColl->size; ++i) {
+		PyrSlot argSlot;
+		getIndexedSlot(argsColl, &argSlot, i);
+		if (!isKindOfSlot(&argSlot, class_string))
+			return errWrongType;
+		char *arg = new char[slotRawObject(&argSlot)->size + 1];
```

Comment:
this is still a leak if you exit early out of this for loop

Reply:
I also think you should move this loop out into a separate static method so that it's easier to understand. it could fill the vector with args and return an empty-vector on failure. this would mean separating the `new char[]` from the `delete []`.. maybe it would be safer to create a `vector<string>` which you then use to fill a `char* []`, then there is no explicit memory management required. it's fine as it is (minus memory leak) but would give you a lot of what you're going for, just more safely and with less effort

Reply:
it would also mean you can factor both these loops out of prPipeOpenArgv and prArrayPOpen!

Reply:
possible signature - `int PyrCollToStrings(PyrObject [const]* coll, std::vector<std::string>& strings)`, using the return type as an error code.